Correct shard disconnect.

islemaster opened the pull request:
### Correct duplicate PubSub subscriptions on shard reset

Adds a proper disconnect/unsubscribe step for the shard, which is triggered when navigating away or when the shard is reset.

Before we kept the shard around unchanged, and just replaced it on reconnect (an artifact of an old attempt at an auto-reconnect system).

This didn't work anymore because our PubSub subscription steps are tied to shard creation, and we need to unsubscribe properly so we don't build up multiple subscriptions across shard resets.

Now we properly unsubscribe/resubscribe PubSub (and use some previously unused code to this effect, go figure).
### Fix logic for showing shard reset button to admins

There was a bug that hid the shard reset button from admins if the shard ID did not include a numeric section ID on the end.  Fixed and brought under test.

Completes parts of [Pivotal #101529076](https://www.pivotaltracker.com/story/show/101529076).

Hamms commented on apps/src/netsim/NetSimTable.js:
```diff
@@ -217,8 +210,7 @@ NetSimTable.prototype.subscribe = function () {
 NetSimTable.prototype.unsubscribe = function () {
-  this.channel_.unsubscribe(this.tableName_, this.channelCallback_);
-  this.channelCallback = undefined;
+  this.channel_.unsubscribe(this.tableName_);
```
Could you update the comments to reflect the minor functionality change here? Also, given that we're no longer specifying which callback we're unsubscribing, it would be nice to make explicit that we're unsubscribing everything

islemaster replied:
:/ We are technically unsubscribing the last callback associated with the `tableName_` key.  It's unsafe to subscribe multiple callbacks with the same key now - I guess that's not great, it just reflects the way we want to use it for now.

Hamms replied:
oh right! I got lost in the hierarchy; too many things with `.unsubscribe` methods.

islemaster replied:
Did update comments to remove references to 'saved callback' since it's no longer a concern at this layer; other issues addressed in PubSubChannel.

Hamms commented on apps/src/netsim/NetSimShard.js:
```diff
@@ -119,6 +128,20 @@ var NetSimShard = module.exports = function (shardID, pubSubConfig) {
+  this.wireTable.unsubscribe();
+  this.messageTable.unsubscribe();
+  this.logTable.unsubscribe();
+  this.pubSubChannel.unsubscribe(ALL_TABLES);
```
It would be nice to making the naming a little more clear here; I could easily see there being some future confusion about the `ALL_TABLES` subscription being for events that span all the tables as opposed to all the events for all the tables.

islemaster replied:
Renamed ALL_TABLES => WHOLE_SHARD_EVENT.

Hamms commented on apps/src/netsim/PubSubChannel.js:
```diff
@@ -61,27 +61,36 @@ PubSubChannel.NullChannel.prototype.unsubscribe = function (eventName, callback)
  */
 PubSubChannel.PusherChannel.prototype.subscribe = function (eventName, callback) {
   this.pusherChannel_.bind(eventName, callback);
+  this.callbacks_[eventName] = callback;
```
Do we want to complain here if this call is going to overwrite the cached callback? Otherwise, we lose the reference to it and can't cleanly unsubscribe.

islemaster replied:
Ah, you already saw this.  Sure, let's complain if this happens.

Hamms replied:
Alternatively, we could unsubscribe the old one before subscribing the new one, but that functionality is a little harder to make obvious

islemaster replied:
Throws Error on double-subscribe.

Hamms commented:
first pass done! Left some questions for ya.

islemaster commented:
@Hamms please take another look.  Addressed comments, and added a fix for a bug that hid the shard reset button from admins if the shard ID did not include a numeric section ID on the end.  Fixed and brought under test.

Hamms commented:
LGTM!
